function/avro-import.py

```diff
@@ -0,0 +1,156 @@
+        write_to_bucket(upsert_file_name, upsert_json_str)
+    except Exception as e:
+        return handle_exception(job_id, start_time, "The following exception occurred: " + str(e))
+    except:
+        return handle_exception(job_id, start_time, "Something went wrong with the following request: " + str(request.json))
+
+
+def handle_exception(subdirectory_name, start_time, message):
+    error_file_name = subdirectory_name + "/error-" + start_time + ".txt"
+    write_to_bucket(error_file_name, message)
+    return message
+
+
+def write_upsert_to_bucket(job_id, content_string):
+    compressed_value = gzip.compress(bytes(content_string, 'utf-8'))
+    file_name = job_id + "/upsert.json"
+    write_to_bucket(file_name, compressed_value)
+
+
+def write_metadata_to_bucket(job_id, content_string):
+    file_name = job_id + "/metadata.json"
+    write_to_bucket(file_name, content_string)
+
+
+def write_to_bucket(file_name, content):
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket("avro-translated-json")
+    blob = bucket.blob(file_name)
+    blob.upload_from_string(content)
+
+
+def translate(reader, options=None):
+    if options is None:
+        options = {}
+    t = Translator(options)
+    return t.translate(reader)
+
+
+class Translator:
+    def __init__(self, options=None):
+        if options is None:
+            options = {}
+        defaults = {'b64-decode-enums': False, 'prefix-object-ids': False}
+        self.options = {**defaults, **options}
+
+    def translate(self, reader):
+        if reader is None:
+            return None
+
+        enums = _list_enums(reader.writer_schema)
+        results = [self._translate_record(record, enums)
+                   for record in reader if record['name'] != 'Metadata']
+        return results
+
+    def _translate_record(self, record, enums):
+        entity_type = record['name']
+        name = record['id']
+
+        def make_op(key, value):
+            if self.options['b64-decode-enums'] and (entity_type, key) in enums:
+                value = _b64_decode(value).decode("utf-8")
+            if self.options['prefix-object-ids'] and key == 'object_id':
+                value = 'drs://' + value
+            return _make_add_update_op(key, value)
+
+        attributes = [make_op(key, value)
+                      for key, value in record['object'].items() if value is not None]
+        relations = [make_op(relation['dst_name'],
+                             {'entityType': relation['dst_name'], 'entityName': relation['dst_id']})
+                     for relation in record['relations']]
+
+        return {
+            'name': name,
+            'entityType': entity_type,
+            'operations': [*attributes, *relations]
+        }
+
+
+def _b64_decode(encoded_value):
+    return base64.b64decode(encoded_value + "=" * (-len(encoded_value) % 4))
+
+
+def _list_enums(schema):
+    object_field = next(f for f in schema['fields'] if f['name'] == 'object')
```

Comment:
this is generating a tuple of N elements, and then pulling the first element off the tuple and throwing the rest away. is this an implementation of `first`, similar to [this post](https://stackoverflow.com/questions/2361426/get-the-first-item-from-an-iterable-that-matches-a-condition)? if so i would recommend actually making the function and calling it `first` because i had to do some mental gymnastics to figure out what was going on here

Reply:
The intent here is more `only` than `first`. The schema has fields, only one of which is named "object". It's a good point. I've definitely done this before... written a function called `only` that pulls the first item from an iterator... to clarify the intent. I would be in favor of doing that here.